text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,764 @@
+[unresolved-questions]: #unresolved-questions
+
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+    * Whether to pick an alternative syntax (and which one in that case).
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+    * We've already handled this since the last RFC, there are no more implementation concerns.
+- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?
+    * This RFC's syntax is entirely unrelated to discussions on `async Trait`.
+        * `async Trait` can be written entirely in user code by creating a new trait `AsyncTrait`; there is no workaround for `const`.
+    * This RFC's syntax is entirely unrelated to discussions on effect syntax.
+        * If we get an effect system, it may be desirable to allow expressing const traits with the effect syntax, this design is forward compatible with that.
+        * If we get an effect system, we will still want this shorthand, just like we allow you to write:
+            * `T: Iterator<Item = U>` and don't require `where T: Iterator, <T as Iterator>::Item = U`.
+            * `T: Iterator<Item: Debug>` and don't require `where T: Iterator, <T as Iterator>::Item: Debug`.
+    * RTN for per-method bounds: `T: Trait<some_fn(..): ~const Fn(A, B) -> C>` could supplement this feature in the future.
+        * Alternatively `where <T as Trait>::some_fn(..): ~const` or `where <T as Trait>::some_fn \ {const}`.
+        * Very verbose (need to specify arguments and return type).
+        * Want short hand sugar anyway to make it trivial to change a normal function to a const function by just adding some minor annotations.
+        * Significantly would delay const trait stabilization (by years).
+        * Usually requires editing the trait anyway, so there's no "can constify impls without trait author opt in" silver bullet.
+    * New RTN-like per-method bounds: `T: Trait<some_fn(_): ~const>`.
+        * Unclear if soundly possible.
+        * Unclear if possible without incurring significant performance issues for all code (may need tracking new information for all functions out there).
+        * Still requires editing traits.
+        * Still want the `~const Trait` sugar anyway.
+
+## Should we start out by allowing only const trait declarations and const trait impls
+
+We do not need to immediately allow using methods on generic parameters of const fn, as a lot of const code is nongeneric.
+
+The following example could be made to work with just const traits and const trait impls.
+
+```rust
+const fn foo() {
+    let a = [1, 2, 3];
+    let b = [1, 2, 4];
+    if a == b {}
+}
+```
+
+Things like `Option::map` could not be made const without const trait bounds, as they need to actually call the generic `FnOnce` argument.
+
+
+# Future possibilities
```

Comment:
Should we mention closures here. CC https://github.com/rust-lang/project-const-traits/issues/10